hil/ext/switches/dellnos9.py

```diff
@@ -0,0 +1,380 @@
+        except ValueError:
+            # happens when our response has almost no info on a switchport
+            # because it's off. the response lacks the word "vlanmembership"
+            return []
+        if response[index] == '':
+            # a port with no tagged interface has no section with "T"
+            return []
+        vlan_list = response[index].replace('T', '').split(',')
+        return [('vlan/%s' % x, x) for x in vlan_list]
+
+    def _get_native_vlan(self, interface):
+        """ Return the native vlan of an interface.
+
+        Args:
+            interface: interface to return the native vlan of
+
+        Returns: Tuple of the form ('vlan/native', vlan) or None
+
+        Similar to _get_vlans()
+        """
+        response = self._get_port_info(interface)
+        if response.find('NativeVlanId:') == -1:
+            return None
+        begin = response.find('NativeVlanId:') + len('NativeVlanId:')
+        end = response.find('.', begin)
+        vlan = response[begin:end]
+
+        # FIXME: At this point, we are unable to remove the default native vlan
+        # The switchport defaults to native vlan 1. Our deployment tests make
+        # assertions that a switchport has no default vlan at start.
+        # To get the tests to pass, we return None if we see the switchport has
+        # a default native vlan (=1). This needs to be fixed ASAP.
+        if vlan == '1':
+            return None
+
+        return ('vlan/native', vlan)
+
+    def _get_port_info(self, interface):
+        """Returns the output of a show interface command. This removes all
+        spaces from the response before returning it which is then parsed by
+        the caller.
+
+        Sample Response:
+        u"<outputxmlns='http://www.dell.com/ns/dell:0.1/root'>\n
+        <command>show interfaces switchport GigabitEthernet1/3\r\n\r\n
+        Codes: U-Untagged T-Tagged\r\n x-Dot1x untagged,X-Dot1xtagged\r\n
+        G-GVRP tagged,M-Trunk\r\n i-Internal untagged, I-Internaltagged,
+        v-VLTuntagged, V-VLTtagged\r\n\r\n Name:GigabitEthernet1/3\r\n 802.1Q
+        Tagged:Hybrid\r\n Vlan membership:\r\n Q Vlans\r\n U 1512 \r\n T 1511
+        \r\n\r\n Native Vlan Id: 1512.\r\n\r\n\r\n\r\n
+        MOC-Dell-S3048-ON#</command>\n</output>\n"
+        """
+
+        command = 'interfaces switchport %s %s' % \
+            (self.interface_type, interface)
+        response = self._execute(interface, SHOW, command)
+        return response.text.replace(' ', '')
+
+    def _add_vlan_to_trunk(self, interface, vlan):
+        """ Add a vlan to a trunk port.
+
+        If the port is not trunked, its mode will be set to trunk.
+
+        Args:
+            interface: interface to add the vlan to
+            vlan: vlan to add
+        """
+        if not self._is_port_on(interface):
+            self._port_on(interface)
+        command = 'interface vlan ' + vlan + '\r\n tagged ' + \
```

Comment:
This driver constructs enough commands with the \r\n line endings that I think it's worthwhile to use a different pattern to reduce the bugs.

Something like this might work well
```
self._build_cmd = lambda cmd: '\r\n'.join(cmd)
```

Reply:
Could you explain how this reduces bugs? I avoid using lambas because they make the code slightly less readable imo.

Reply:
sorry to butt in, but please do not assign names to anonymous functions. if you want a named function use def.

Reply:
@jeremyfreudberg feel free to butt in.

Reply:
Yeah, I was just being really lazy. My main point though was to define a function that would build these command sequences with the /r/n instead of constructing them inline. It would take in an array of commands and then return the string delimited by windows newlines.